pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package rafttrace
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/raft"
+	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
+)
+
+type traceValue struct {
+	// ctx is a trace specific context used to log events on this trace.
+	ctx context.Context
+	// baseCtx is the underlying proposal buffer which we additionally log to.
+	baseCtx withContext
+	// cleanup will retrieve all events that were logged in this trace.
+	cleanup func() tracingpb.Recording
+}
+
+// RaftTracer is a utility to trace raft messages. Note that it may erroneously
```

Comment:
nit: Let's not frame it "erroneously". This can be considered a feature, because we see more information about a log index.

An example higher latency case when this additional info might be useful:

- We submit a proposal which get appended as `(term=10,index=100)`, but this was racing with a leader change.
- The new leader at `term=11` has overwritten this entry. We will see this in the trace (there should be a bunch of `MsgApp/MsgStorageAppend` events overlapping this index again).
- This causes the leaseholder to **wait**, and eventually realize that the proposal never popped out committed. It submits a reproposal, which now ends up at `(term=11,index=110)`. Ideally, we also should see that in the trace.

Reply:
I cleaned up the comment and added your example to this case.